File: leetcode/bst/RecoverBinarySearchTree.java
package bst;

public class RecoverBinarySearchTree {
    /* 99
     * 好久不做二叉搜索树的题目完全不知道利用bst的特点
     * 这道题我的想法是遍历然后保持一些up down parent treenode 拿现在的node和这些比较 < down则和down swap
     * 但是一些细节没有考虑好 而且还没有搞清楚swap只要换val就可以了 在那里搞swap node 搞不出来 最后的解法还是有问题
     * 一个简单的答案是inorder traverse 形成list 然后检查颠倒顺序的两个 互换val
     */
    private List<TreeNode> inorder;
    public void recoverTree(TreeNode root) {
        inorder = new ArrayList<>();
        traverse(root);
        int size = inorder.size();
        TreeNode a = null, b = null;
        for(int i = 0;i < size - 1;i++) {
            if(inorder.get(i).val > inorder.get(i+1).val){
                a = inorder.get(i);
                break;
            }
        }
        for(int i = size - 1;i > 0;i--) {
            if(inorder.get(i).val < inorder.get(i-1).val){
                b = inorder.get(i);
                break;
            }
        }
        swap(a, b);
    }
    private void traverse(TreeNode root) {
        if(root == null) return;
        traverse(root.left);
        inorder.add(root);
        traverse(root.right);
    }
    private void swap(TreeNode a, TreeNode b) {
        int temp = a.val;
        a.val = b.val;
        b.val = temp;
    }
    /* 错误答案 2 3 1会把2 3互换 实际答案应该是1 3互换 */
    // private boolean earlyStop = false;
    // public void recoverTree(TreeNode root) {
    //     traverse(root, null, null);
    // }
    // private void traverse(TreeNode root, TreeNode up, TreeNode down) {
    //     if(earlyStop) return;
    //     if(root == null) return;
    //     if(up != null && root.val > up.val) {
    //         swap(root, up);
    //         earlyStop = true;
    //     }else if(down != null && root.val < down.val) {
    //         swap(root, down);
    //         earlyStop = true;
    //     }
    //     traverse(root.left, root, down);
    //     traverse(root.right, up, root);
    // }
}

File: c++/tree/bst/recoverBinarySearchTree.cpp
#include <include/listnode.h>
/**
 * 99. Recover Binary Search Tree
 * 这道题的思路一下子还不好想 我们可以从中序遍历的角度思考 BST用中序遍历展平其实就是升序数列
 * 如何找到一个本来升序的数列中调换的两个数字呢 第一个会比后一个数字大 第二个会比前一个数字小
 * 如果我们将找到第一个作为目标 则第二个又要另一套
 * 这里选择的方法是记录升序数列的前一个数字 一旦发现它比当前大 那么当前在两个位置 一个是第一个错误的后一个 第二个就是在第二个错误
 * 这样我们就更新两个mistake 出来换一下就行了
 * 
 * 比起java版本 这里不用额外的vector来展平BST 
*/
class Solution {
public:
    void recoverTree(TreeNode* root) {
        TreeNode *firstMis = nullptr, *secondMis = nullptr, *prev =nullptr;
        traverse(root, firstMis, secondMis, prev);
        int temp = firstMis->val;
        firstMis->val = secondMis->val;
        secondMis->val = temp;
    }

    void traverse(TreeNode* root, TreeNode*& firstMis, TreeNode*& secondMis, TreeNode*& prev) {
        if(!root) return;
        traverse(root->left, firstMis, secondMis, prev);
        if(prev && root->val < prev->val) {
            if(!firstMis) {
                firstMis = prev;
                secondMis = root;
            }else{
                secondMis = root;
            }
        }
        prev = root;
        traverse(root->right, firstMis, secondMis, prev);
    }
};